### PR TITLE
Fix Content-Lenght calculation of unicode string

### LINKED ISSUE #1482: Too much data for declared Content-Length when passing string with non-ascii characters via files parameter
### Checklist

- [x ] The bug is reproducible against the latest release and/or `master`.
- [ x] There are no similar issues or pull requests to fix it yet.

### Describe the bug

`h11._util.LocalProtocolError: Too much data for declared Content-Length` is raised when passing string with non-ascii characters via files parameter.

### To reproduce

```
import httpx

response = httpx.post(
    "https://httpbin.org/post",
    files={
        'upload-file': ('example.txt', '\u00E9', 'text/plain; charset=utf-8')
    }
)
response.raise_for_status()
print(response.read())
```
-->

### Expected behavior

An exception not be raised and the content length properly computed

### Actual behavior

An exception is raised.

### Debugging material

```
Traceback (most recent call last):
  File "/***redacted***/httpx_bug.py", line 5, in <module>
    response = httpx.post(
  File "/***redacted***/.venv/lib/python3.9/site-packages/httpx/_api.py", line 296, in post
    return request(
  File "/***redacted***/.venv/lib/python3.9/site-packages/httpx/_api.py", line 93, in request
    return client.request(
  File "/***redacted***/.venv/lib/python3.9/site-packages/httpx/_client.py", line 733, in request
    return self.send(
  File "/***redacted***/.venv/lib/python3.9/site-packages/httpx/_client.py", line 767, in send
    response = self._send_handling_auth(
  File "/***redacted***/.venv/lib/python3.9/site-packages/httpx/_client.py", line 805, in _send_handling_auth
    response = self._send_handling_redirects(
  File "/***redacted***/.venv/lib/python3.9/site-packages/httpx/_client.py", line 837, in _send_handling_redirects
    response = self._send_single_request(request, timeout)
  File "/***redacted***/.venv/lib/python3.9/site-packages/httpx/_client.py", line 861, in _send_single_request
    (status_code, headers, stream, ext) = transport.request(
  File "/***redacted***/.venv/lib/python3.9/site-packages/httpcore/_sync/connection_pool.py", line 218, in request
    response = connection.request(
  File "/***redacted***/.venv/lib/python3.9/site-packages/httpcore/_sync/connection.py", line 106, in request
    return self.connection.request(method, url, headers, stream, ext)
  File "/***redacted***/.venv/lib/python3.9/site-packages/httpcore/_sync/http11.py", line 66, in request
    self._send_request_body(stream, timeout)
  File "/***redacted***/.venv/lib/python3.9/site-packages/httpcore/_sync/http11.py", line 112, in _send_request_body
    self._send_event(event, timeout)
  File "/***redacted***/.venv/lib/python3.9/site-packages/httpcore/_sync/http11.py", line 123, in _send_event
    bytes_to_send = self.h11_state.send(event)
  File "/***redacted***/.venv/lib/python3.9/site-packages/h11/_connection.py", line 468, in send
    data_list = self.send_with_data_passthrough(event)
  File "/***redacted***/.venv/lib/python3.9/site-packages/h11/_connection.py", line 501, in send_with_data_passthrough
    writer(event, data_list.append)
  File "/***redacted***/.venv/lib/python3.9/site-packages/h11/_writers.py", line 58, in __call__
    self.send_data(event.data, write)
  File "/***redacted***/.venv/lib/python3.9/site-packages/h11/_writers.py", line 78, in send_data
    raise LocalProtocolError("Too much data for declared Content-Length")
h11._util.LocalProtocolError: Too much data for declared Content-Length
```
-->

### Environment

- OS: macOS
- Python version: `Python 3.9.1`
- HTTPX version: `0.16.1`
- Async environment: n/a
- HTTP proxy: no
- Custom certificates: no

### Additional context
n/a

### PR DESCRIPTION
**Pull request by j178**
Fixes #1482

### CODE REVIEW

**Review by florimondmanca — CHANGES_REQUESTED** [review]
Looking very nice, just a few suggestions to make this tighter?

File: httpx/_multipart.py [hunk 1]
```diff
@@ -74,20 +71,22 @@ def __init__(self, name: str, value: FileTypes) -> None:
...
+        self.file: Union[IO[bytes], IO[str], bytes] = fileobj
         self.content_type = content_type
         self._consumed = False
 
     def get_length(self) -> int:
         headers = self.render_headers()
 
-        if isinstance(self.file, (str, bytes)):
+        if isinstance(self.file, bytes):
             return len(headers) + len(self.file)
```
[the remark this answers is not in the document]
Comment (j178) [reply httpx/_multipart.py:90]: Yes. this is the simple and correct fix, I was trying to avoid an extra `to_bytes` call.

**Comment by tomchristie** [comment]
Looks fiddly, just flagging up that I'd also like to be involved in any review on this. 😇

**Comment by tomchristie** [comment]
Great, thanks for this, good work!

I'd like to see the change footprint reduced on it slightly. I don't think we really need to change `_types.py` at all, and I think we can reduce the number of changes in `_multipart.py`. I'll leave a couple of inline suggestions there.

File: httpx/_multipart.py [hunk 2]
```diff
@@ -40,11 +40,7 @@ def render_headers(self) -> bytes:
 
     def render_data(self) -> bytes:
         if not hasattr(self, "_data"):
-            self._data = (
-                self.value
-                if isinstance(self.value, bytes)
-                else self.value.encode("utf-8")
-            )
```
Comment (tomchristie) [inline httpx/_multipart.py:47]: This is a nice tweak, yup.

File: httpx/_multipart.py [hunk 3]
```diff
@@ -66,8 +62,6 @@ class FileField:
     def __init__(self, name: str, value: FileTypes) -> None:
         self.name = name
 
-        fileobj: FileContent
-
```
Comment (tomchristie) [inline httpx/_multipart.py:70]: I'd suggest we don't change this.

File: httpx/_multipart.py [hunk 4]
```diff
@@ -79,6 +73,12 @@ def __init__(self, name: str, value: FileTypes) -> None:
             fileobj = value
             content_type = guess_content_type(filename)
 
+        if isinstance(fileobj, str):
+            # Ensure we only deal with bytes to prevent any content-length
+            # mismatch due to str -> bytes encoding.
+            # See: https://github.com/encode/httpx/issues/1482
+            fileobj = to_bytes(fileobj)
+
```
Comment (tomchristie) [inline httpx/_multipart.py:81]: Or add this.

File: httpx/_multipart.py [hunk 5]
```diff
@@ -87,7 +87,7 @@ def __init__(self, name: str, value: FileTypes) -> None:
     def get_length(self) -> int:
         headers = self.render_headers()
 
-        if isinstance(self.file, (str, bytes)):
```
Comment (tomchristie) [inline httpx/_multipart.py:90]: The low-footprint change here is presumably to keep this check the same, but use the following...

`return len(headers) + len(to_bytes(self.file))`

Right?

File: httpx/_multipart.py [hunk 6]
```diff
@@ -119,8 +119,8 @@ def render_headers(self) -> bytes:
         return self._headers
 
     def render_data(self) -> typing.Iterator[bytes]:
-        if isinstance(self.file, (str, bytes)):
-            yield to_bytes(self.file)
```
Comment (tomchristie) [inline httpx/_multipart.py:123]: We shouldn't need to change this.

**Pull request closed without merging**